### PR TITLE
[GH-#132] Finished integrating the elixir-wallet repository

### LINKED ISSUE #132: replace key worker with elixir-wallet
can be started when https://github.com/aeternity/elixir-wallet/issues/58 is finished

### PR DESCRIPTION
**Pull request by gspasov**
Removed the old Keys module and made a new Wallet module, to handle the elixir-wallet functions. Made a specified path to the `priv dir` inside the `aecore dir`, where the wallets will be stored under `wallets dir`

resolves #132

### CODE REVIEW

File: apps/aecore/lib/aecore/structures/signed_tx.ex [hunk 1]
```diff
@@ -3,33 +3,54 @@ defmodule Aecore.Structures.SignedTx do
   Aecore structure of a signed transaction.
   """
 
-  alias Aecore.Keys.Worker, as: Keys
   alias Aecore.Structures.TxData
   alias Aecore.Structures.SignedTx
 
+  @typedoc "Structure of the TxData"
+  @type tx() :: %TxData{}
```
Comment (thepiwo) [inline apps/aecore/lib/aecore/structures/signed_tx.ex]: this should be `@type t() :: %TxData{}` in tx_data.ex file

File: apps/aecore/lib/aecore/structures/signed_tx.ex [hunk 2]
```diff
@@ -3,33 +3,54 @@ defmodule Aecore.Structures.SignedTx do
...
+
+  @typedoc "Structure of the SignedTx module"
   @type t :: %SignedTx{
-    data: TxData.t(),
+    data: tx(),
```
Comment (thepiwo) [inline apps/aecore/lib/aecore/structures/signed_tx.ex]: keep this as `TxData.t()`

File: apps/aecore/lib/aecore/structures/signed_tx.ex [hunk 3]
```diff
@@ -3,33 +3,54 @@ defmodule Aecore.Structures.SignedTx do
...
+     * data: Aecore %TxData{} structure
+     * signature: Signed %TxData{} with the private key of the sender
   """
   defstruct [:data, :signature]
   use ExConstructor
 
-  @spec is_coinbase?(SignedTx.t()) :: boolean()
-  def is_coinbase?(tx) do
-    tx.data.from_acc == nil && tx.signature == nil
+  @spec is_coinbase(signed_tx()) :: boolean()
```
Comment (thepiwo) [inline apps/aecore/lib/aecore/structures/signed_tx.ex]: keep this as `SignedTx.t()`

File: apps/aecore/lib/aecore/structures/signed_tx.ex [hunk 4]
```diff
@@ -3,33 +3,54 @@ defmodule Aecore.Structures.SignedTx do
...
+  def is_coinbase(%{data: %{from_acc: key}, signature: signature}) do
+    key == nil && signature == nil
+  end
+
+  @spec is_valid(signed_tx()) :: boolean()
```
Comment (thepiwo) [inline apps/aecore/lib/aecore/structures/signed_tx.ex]: keep this as `SignedTx.t()`

File: apps/aecore/lib/aecore/structures/signed_tx.ex [hunk 5]
```diff
@@ -3,33 +3,54 @@ defmodule Aecore.Structures.SignedTx do
...
+  def is_valid(%{data: data, signature: signature}) do
+    data.value >= 0 && data.fee >= 0
+      Aewallet.Signing.verify(:erlang.term_to_binary(data), signature, data.from_acc)
```
Comment (thepiwo) [inline apps/aecore/lib/aecore/structures/signed_tx.ex]: is there an `&&` missing? otherwise, how does the verification work?

File: apps/aecore/lib/aecore/structures/signed_tx.ex [hunk 6]
```diff
@@ -3,33 +3,54 @@ defmodule Aecore.Structures.SignedTx do
...
+  Takes the transaction that needs to be signed
+  and the private key of the sender.
+  Returns a signed tx
+
+  ## Parameters
+     - tx: The transaction data that it's going to be signed
+     - priv_key: The priv key to sign with
+
+  """
+  @spec sign_tx(tx(), binary()) :: {:ok, %SignedTx{}}
```
Comment (thepiwo) [inline apps/aecore/lib/aecore/structures/signed_tx.ex]: `@spec sign_tx(TxData.t(), binary()) :: {:ok, SignedTx.t()}`

File: apps/aecore/lib/aecore/wallet/worker.ex [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+  Module for handling the creation of a Wallet file
+  """
+
+  use GenServer
+
+  @typedoc "Options for network"
+  @type opts :: :mainnet | :testnet
+
+  @aewallet_dir Application.get_env(:aecore, :aewallet)[:path]
+  @aewallet_pass "1234"
```
Comment (thepiwo) [inline apps/aecore/lib/aecore/wallet/worker.ex]: why do we choose "1234" as password over "" as default?

File: apps/aecore/lib/aecore/miner/worker.ex [hunk 8]
```diff
@@ -12,18 +12,19 @@ defmodule Aecore.Miner.Worker do
...
   alias Aecore.Txs.Pool.Worker, as: Pool
   alias Aeutil.Bits
   alias Aecore.Peers.Worker, as: Peers
+  alias Aecore.Wallet.Worker, as: Wallet
 
   require Logger
 
   @mersenne_prime 2147483647
   @coinbase_transaction_value 100
+  @wallet_default_pass "1234"
```
Comment (thepiwo) [inline apps/aecore/lib/aecore/miner/worker.ex]: why do we choose "1234" as password over "" as default?

File: apps/aecore/lib/aecore/wallet/worker.ex [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+    end
+  end
+
+  defp create_wallet(path) do
+    Aewallet.Wallet.create_wallet(@aewallet_pass, path)
+  end
+
+  defp get_file_name(path) do
+    path
+    |> Path.join("*/")
```
Comment (thepiwo) [inline apps/aecore/lib/aecore/wallet/worker.ex]: Why is this needed?

File: apps/aecore/lib/aecore/miner/worker.ex (hunk 8 above)
Comment (gspasov) [reply apps/aecore/lib/aecore/miner/worker.ex]: Hmm I just did thought of that, I'll change it

File: apps/aecore/lib/aecore/structures/signed_tx.ex (hunk 5 above)
Comment (gspasov) [reply apps/aecore/lib/aecore/structures/signed_tx.ex]: Yes, my fault. I saw it later, it's fixed

File: apps/aecore/lib/aecore/wallet/worker.ex (hunk 7 above)
Comment (gspasov) [reply apps/aecore/lib/aecore/wallet/worker.ex]: As I mentioned up, I'll fix this

File: apps/aecore/lib/aecore/wallet/worker.ex (hunk 9 above)
Comment (gspasov) [reply apps/aecore/lib/aecore/wallet/worker.ex]: This is done so that I could search for files I'm not sure if exist inside the aewallet dir. This star could be actually added inside the config path. I'll fix it.

File: apps/aecore/lib/aecore/wallet/worker.ex [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+    case get_file_name(@aewallet_dir) do
+      []  -> create_wallet(@aewallet_dir)
+      [_] -> :ok
+    end
+  end
+  defp has_wallet(false) do
+    if File.mkdir @aewallet_dir do
+      create_wallet(@aewallet_dir)
+    else
+      throw("Failed creating the wallet dir: #{@aewallet_dir}")
```
Comment (d-velev) [inline apps/aecore/lib/aecore/wallet/worker.ex]: this should be caught in the init

File: apps/aecore/lib/aecore/wallet/worker.ex [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+  @typedoc "Options for network"
+  @type opts :: :mainnet | :testnet
+
+  @aewallet_dir Application.get_env(:aecore, :aewallet)[:path]
+  @aewallet_pass "1234"
+
+  ## Client API
+
+  def start_link(_args) do
+    GenServer.start_link(__MODULE__, %{path: @aewallet_dir}, name: __MODULE__)
```
Comment (d-velev) [inline apps/aecore/lib/aecore/wallet/worker.ex]: Why do we need a GenServer when we don't need to keep state?

File: apps/aecore/lib/aecore/miner/worker.ex [hunk 12]
```diff
@@ -233,7 +234,7 @@ defmodule Aecore.Miner.Worker do
       valid_txs_by_chainstate = BlockValidation.filter_invalid_transactions_chainstate(ordered_txs_list, chain_state, top_block.header.height + 1)
       valid_txs_by_fee = filter_transactions_by_fee(valid_txs_by_chainstate)
 
-      {_, pubkey} = Keys.pubkey()
+      pubkey = Wallet.get_public_key(@wallet_default_pass)
```
Comment (d-velev) [inline apps/aecore/lib/aecore/miner/worker.ex]: What if the password is changed?

File: apps/aecore/lib/aecore/miner/worker.ex (hunk 8 above)
Comment (d-velev) [inline apps/aecore/lib/aecore/miner/worker.ex]: If this is used in multiple locations then maybe it should be stored in the config or have it as an attribute in one module and have a getter function for it.

File: apps/aecore/lib/aecore/structures/signed_tx.ex [hunk 13]
```diff
@@ -46,9 +47,9 @@ defmodule Aecore.Structures.SignedTx do
      - priv_key: The priv key to sign with
 
   """
-  @spec sign_tx(tx(), binary()) :: {:ok, %SignedTx{}}
+  @spec sign_tx(tx(), binary()) :: {:ok, t()}
```
Comment (thepiwo) [inline apps/aecore/lib/aecore/structures/signed_tx.ex]: Is it possible to write `SignedTx.t()` instead of just `t()`?

File: apps/aecore/lib/aecore/wallet/worker.ex [hunk 14]
```diff
@@ -0,0 +1,84 @@
...
+    {:ok, state}
+  end
+
+  @spec get_public_key(String.t()) :: binary()
+  def get_public_key(password) do
+    GenServer.call(__MODULE__, {:get_pub_key, {password, :mainnet}})
+  end
+
+  @spec get_public_key(String.t(), opts()) :: binary()
+  def get_public_key(password, network) do
```
Comment (d-velev) [inline apps/aecore/lib/aecore/wallet/worker.ex]: Do we need to protect the public key with a password?

File: apps/aecore/lib/aecore/structures/signed_tx.ex [hunk 15]
```diff
@@ -3,33 +3,55 @@ defmodule Aecore.Structures.SignedTx do
...
+
+  ## Parameters
+     - tx: The transaction data that it's going to be signed
+     - priv_key: The priv key to sign with
+
+  """
+  @spec sign_tx(tx(), binary()) :: {:ok, t()}
+  def sign_tx(tx, priv_key) do
+    signature = Signing.sign(:erlang.term_to_binary(tx), priv_key)
+    {:ok, %SignedTx{data: tx, signature: signature}}
```
Comment (d-velev) [inline apps/aecore/lib/aecore/structures/signed_tx.ex]: Maybe this should just return the SignedTx without :ok since there are no checks if this is the desired result.

File: apps/aecore/lib/aecore/miner/worker.ex (hunk 8 above)
Comment (cheezus1) [reply apps/aecore/lib/aecore/miner/worker.ex]: Maybe store it in the config, as it is used both in the code and the tests.

File: apps/aecore/lib/aecore/wallet/worker.ex (hunk 14 above)
Comment (gspasov) [reply apps/aecore/lib/aecore/wallet/worker.ex]: The password is used to decrypt the wallet. Everytime we derive some key, we need to decrypt the wallet.

File: apps/aecore/lib/aecore/wallet/worker.ex (hunk 11 above)
Comment (gspasov) [reply apps/aecore/lib/aecore/wallet/worker.ex]: The GenServer was used so that we can start it from the Supervisor because we need to be sure that a wallet is created before starting the miner.

File: apps/aecore/lib/aecore/structures/signed_tx.ex (hunk 15 above)
Comment (gspasov) [reply apps/aecore/lib/aecore/structures/signed_tx.ex]: Well, I'll keep it like this and make some verifications, as well in the TxData module, because now when you mention it I can see that it's needed.

File: apps/aecore/lib/aecore/miner/worker.ex (hunk 8 above)
Comment (gspasov) [reply apps/aecore/lib/aecore/miner/worker.ex]: Yup, on it.

File: apps/aecore/lib/aecore/miner/worker.ex (hunk 12 above)
Comment (gspasov) [reply apps/aecore/lib/aecore/miner/worker.ex]: Same as the above question. I'm on it.

File: apps/aecore/lib/aecore/wallet/worker.ex (hunk 10 above)
Comment (gspasov) [reply apps/aecore/lib/aecore/wallet/worker.ex]: Well in the init I check if there is a Wallet file and then call the create_wallet func, but I'll think about this a bit more.

File: apps/aecore/lib/aecore/wallet/worker.ex (hunk 11 above)
Comment (d-velev) [reply apps/aecore/lib/aecore/wallet/worker.ex]: You can call the function that does this in the aecore.ex module (the start function) and get rid of the GenServer.

Comment (gspasov) [reply apps/aecore/lib/aecore/wallet/worker.ex]: Well now when I think about it, the state could store the public key of the wallet, so that the module won't need to derive it everytime it is needed.

Comment (thepiwo) [reply apps/aecore/lib/aecore/wallet/worker.ex]: @gspasov I think so as well

File: apps/aecore/lib/aecore/wallet/worker.ex [hunk 16]
```diff
@@ -0,0 +1,92 @@
...
+
+  use GenServer
+
+  alias Aewallet.Wallet
+
+  @typedoc "Options for network"
+  @type opts :: :mainnet | :testnet
+
+  @aewallet_dir Application.get_env(:aecore, :aewallet)[:path]
+  @aewallet_pass Application.get_env(:aecore, :aewallet)[:pass]
```
Comment (cytadela8) [inline apps/aecore/lib/aecore/wallet/worker.ex]: If I'm not mistaken this will get evaluated compile time, not runtime!

File: apps/aecore/lib/aecore/wallet/worker.ex [hunk 17]
```diff
@@ -0,0 +1,92 @@
...
+  end
+
+  @spec get_private_key(String.t(), opts()) :: binary()
+  def get_private_key(password, network) do
+    GenServer.call(__MODULE__, {:get_priv_key, {password, network}})
+  end
+
+  ## Server Callbacks
+
+  def handle_call({:get_pub_key, {password, network}}, _from, %{pubkey: nil} = state) do
```
Comment (cytadela8) [inline apps/aecore/lib/aecore/wallet/worker.ex]: I think this genserver could also handle "password" and "network" parameters internally. (store them, probably load them from config)

Comment (gspasov) [reply apps/aecore/lib/aecore/wallet/worker.ex]: Having them stored initialy actually loses the purpose of having them in the first place. In my opinion a password should not be stored anywhere and the network should be able to change on the fly.

File: apps/aecore/lib/aecore/wallet/worker.ex (hunk 16 above)
Comment (gspasov) [reply apps/aecore/lib/aecore/wallet/worker.ex]: Will make two function for calling them at run time

**Comment by thepiwo** [comment]
@meivantodorov can you review?

File: apps/aecore/lib/aecore/wallet/worker.ex (hunk 17 above)
Comment (cytadela8) [reply apps/aecore/lib/aecore/wallet/worker.ex]: - I think the network should be part of some config, becouse changing the network requires preaty much to restart everything (different database for blocks, different genesis block, different peers). Also even if we would want to change it on the fly it still would be nice if the information would be stored in the GenServer, this way code that wants to sign something doesn't have to pass the information about network every time. 
- By storing password I mean storing in the state of GenServer in RAM.

File: apps/aecore/lib/aecore/miner/worker.ex [hunk 18]
```diff
@@ -233,7 +233,7 @@ defmodule Aecore.Miner.Worker do
       valid_txs_by_chainstate = BlockValidation.filter_invalid_transactions_chainstate(ordered_txs_list, chain_state, top_block.header.height + 1)
       valid_txs_by_fee = filter_transactions_by_fee(valid_txs_by_chainstate)
 
-      {_, pubkey} = Keys.pubkey()
+      pubkey = Wallet.get_public_key(Wallet.get_aewallet_dir())
```
Comment (meivantodorov) [inline apps/aecore/lib/aecore/miner/worker.ex]: I think if you are calling internal function to get the path then better do it inside the Wallet.

File: apps/aecore/lib/aecore/wallet/worker.ex (hunk 17 above)
Comment (meivantodorov) [reply apps/aecore/lib/aecore/wallet/worker.ex]: May be we can let the user to decide if he/she wants to keep the password into the memory or to type it every time.

**Comment by thepiwo** [comment]
can you add to the readme that `build-essential autoconf autogen libtool libgmp3-dev` are dependencies

**Comment by thepiwo** [comment]
@nedelcho-delchev-tues @cytadela8 @d-velev @meivantodorov @cheezus1 can you review?

**Comment by d-velev** [comment]
Can you replace the :erlang.term_to_binary with Serialization.pack_binary like it was before?

File: apps/aecore/lib/aecore/wallet/worker.ex [hunk 19]
```diff
@@ -0,0 +1,143 @@
...
+    {:reply, key, state}
+  end
+
+  def handle_call({:get_priv_key, {password, network}}, _from, %{path: path} = state) do
+    {:ok, priv_key} =
+      Wallet.get_private_key(get_file_name(path), password, network: network)
+    {:reply, priv_key, state}
+  end
+
+  def handle_call({:bech32_encode, {pub_key, prefix}}, _from, state) do
```
Comment (d-velev) [inline apps/aecore/lib/aecore/wallet/worker.ex]: No reason for this or the decode function to be calls.

File: apps/aecore/test/multiple_transactions_test.exs [hunk 20]
```diff
@@ -279,172 +301,183 @@ defmodule MultipleTransactionsTest do
...
   end
 
   defp create_signed_tx(from_acc, to_acc, value, nonce, fee, lock_time_block \\ 0) do
     {from_acc_pub_key, from_acc_priv_key} = from_acc
     {to_acc_pub_key, _to_acc_priv_key} = to_acc
     {:ok, tx_data} = SpendTx.create(from_acc_pub_key, to_acc_pub_key, value,
-                                   nonce, fee, lock_time_block)
-    {:ok, signature} = Keys.sign(tx_data, from_acc_priv_key)
+      nonce, fee, lock_time_block)
+    {:ok, %{signature: signature}} = SignedTx.sign_tx(tx_data, from_acc_priv_key)
```
Comment (cytadela8) [inline apps/aecore/test/multiple_transactions_test.exs]: Here you are extracting the signature from SignedTx just to create the signed_tx again. You should instead:
```
    {:ok, signed_tx} = SignedTx.sign_tx(tx_data, from_acc_priv_key)
    signed_tx
end
```

**Comment by thepiwo** [comment]
@cheezus1 @d-velev can you re-review?

**Comment by thepiwo** [comment]
@cheezus1 @d-velev can you re-review?

File: apps/aecore/lib/aecore/structures/signed_tx.ex [hunk 21]
```diff
@@ -24,13 +25,33 @@ defmodule Aecore.Structures.SignedTx do
...
+  def is_coinbase?(%{data: %{from_acc: key}, signature: signature}) do
+    key == nil && signature == nil
   end
 
   @spec is_valid?(SignedTx.t()) :: boolean()
-  def is_valid?(tx) do
-    tx.data.value >= 0 && tx.data.fee >= 0 && Keys.verify_tx(tx)
+  def is_valid?(%{data: data, signature: signature}) do
+    data.value >= 0 && data.fee >= 0 &&
+      Signing.verify(Serialization.pack_binary(data), signature, data.from_acc)
```
Comment (d-velev) [inline apps/aecore/lib/aecore/structures/signed_tx.ex]: Maybe move pack_binary to Signing.verify, it would be easier to change it and it won't have to be written every time something is signed

File: apps/aecore/lib/aecore/wallet/worker.ex [hunk 22]
```diff
@@ -0,0 +1,155 @@
...
+  @typedoc "Wallet type"
+  @type wallet_type :: :ae | :btc
+
+  @typedoc "Options for network"
+  @type opts :: :mainnet | :testnet
+
+  ## Client API
+
+  def start_link(_args) do
+    GenServer.start_link(__MODULE__, %{path: get_aewallet_dir(), pubkey: nil}, name: __MODULE__)
```
Comment (d-velev) [inline apps/aecore/lib/aecore/wallet/worker.ex]: I think the path shouldn't be stored here since it's already in the env and there is no function to alter it. In general, I think that functions that change the state should be available if a genserver is used.

File: apps/aecore/lib/aecore/structures/signed_tx.ex (hunk 21 above)
Comment (gspasov) [reply apps/aecore/lib/aecore/structures/signed_tx.ex]: This function is implemented in the `elxir-wallet` repository so I don't think it's a good idea to change it's base functionality for this. It takes binary as the first argument and I'd like to keep it this way. This way the function is more generic and not bouned to some specific implementation.

Comment (thepiwo) [reply apps/aecore/lib/aecore/structures/signed_tx.ex]: ok, lets keep it like this

File: apps/aecore/lib/aecore/wallet/worker.ex (hunk 22 above)
Comment (thepiwo) [reply apps/aecore/lib/aecore/wallet/worker.ex]: @d-velev can you approve now?

**Comment by d-velev** [comment]
Is it possible to create a new keypair for testing purposes? For example: we already have a wallet file created and it is read by 2 nodes on one machine, if they both have the same public key we would want to create a new one on one of the nodes so that we can see chainstate changes.

**Comment by thepiwo** [comment]
@gspasov can you give an answer here, so we can merge soon?

**Comment by gspasov** [comment]
@thepiwo I'm going to work on the comment/suggestion that @d-velev gave me.
Sorry for the delay

File: apps/aecore/lib/aecore/wallet/worker.ex [hunk 23]
```diff
@@ -47,32 +48,42 @@ defmodule Aecore.Wallet.Worker do
 
   @spec get_public_key() :: binary()
   def get_public_key() do
-    get_public_key(get_aewallet_pass())
+    get_public_key("")
```
Comment (thepiwo) [inline apps/aecore/lib/aecore/wallet/worker.ex]: why is the default password hardcoded now?

File: apps/aecore/lib/aecore/structures/signed_tx.ex [hunk 24]
```diff
@@ -25,13 +26,33 @@ defmodule Aecore.Structures.SignedTx do
...
+  and the private key of the sender.
+  Returns a signed tx
+
+  ## Parameters
+     - tx: The transaction data that it's going to be signed
+     - priv_key: The priv key to sign with
+
+  """
+  @spec sign_tx(SpendTx.t(), binary()) :: {:ok, SignedTx.t()}
+  def sign_tx(%SpendTx{} = tx, priv_key) when byte_size(priv_key) == 32 do
```
Comment (thepiwo) [inline apps/aecore/lib/aecore/structures/signed_tx.ex]: is it possible to not pass the priv_key here, but have it in the `Signing.sign` method?

File: apps/aecore/lib/aecore/wallet/worker.ex (hunk 23 above)
Comment (gspasov) [reply apps/aecore/lib/aecore/wallet/worker.ex]: No, it's not hardcoded, this is the default `derivation_path`. Now as first argument to the `get_public_key()` I have the `derivation_path`, so the full func is `get_public_key(derivation_path, passoword, network)`

Comment (thepiwo) [reply apps/aecore/lib/aecore/wallet/worker.ex]: okay, can we make a variable for the default `derivation_path` then?

File: apps/aecore/lib/aecore/structures/signed_tx.ex (hunk 24 above)
Comment (gspasov) [reply apps/aecore/lib/aecore/structures/signed_tx.ex]: No, I need the private key to be passed to the `Aewallet.Signing.sing()` function, otherwise it cannot sign the data

Comment (thepiwo) [reply apps/aecore/lib/aecore/structures/signed_tx.ex]: can you fetch it in `sign_tx`?

Comment (gspasov) [reply apps/aecore/lib/aecore/structures/signed_tx.ex]: Well, I guess I could call `Wallet.get_private_key()`

Comment (thepiwo) [reply apps/aecore/lib/aecore/structures/signed_tx.ex]: ok, would be better than passing the private_key every time

Comment (thepiwo) [reply apps/aecore/lib/aecore/structures/signed_tx.ex]: can't this be called in the `Signing.sign` function?

Comment (gspasov) [reply apps/aecore/lib/aecore/structures/signed_tx.ex]: This is the function in the `elixir-wallet` repo

Comment (thepiwo) [reply apps/aecore/lib/aecore/structures/signed_tx.ex]: ok, thanks, I didn't notice

**Comment by gspasov** [comment]
@thepiwo can you check and merge?

File: apps/aecore/lib/aecore/structures/signed_tx.ex (hunk 24 above)
Comment (thepiwo) [reply apps/aecore/lib/aecore/structures/signed_tx.ex]: @gspasov the private-ley is still passed, could you pass `Wallet.get_private_key()` as mentioned?
